### ISSUE TITLE
Static autoloader is not working when project directory is /{vendor-name}

### ISSUE THREAD

**Issue by kachkaev** [composer/composer#5564]
I was playing with Docker and wanted to make a simple cli-only app on top of php 5.6 & Symfony 3.1 – nothing specific in `composer.yml`. `composer install` was working perfectly on my host machine while the project was in `/path/to/project` directory.

Something broke when I decided to move the app into a docker container, mounting `/path/to/project` to `/symfony` (a folder that is located right in `/`). Running `composer install` from within docker started causing a pretty strange error:

```
post-install-cmd: Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::buildBootstrap

Fatal error: Class 'Symfony\Component\ClassLoader\ClassCollectionLoader' not found
in /symfony/vendor/sensio/distribution-bundle/Composer/ScriptHandler.php on line 282
Script Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::buildBootstrap
handling the post-install-cmd event terminated with an exception

[RuntimeException]                                     
An error occurred when generating the bootstrap file.  
``` 

My first guess was that something was wrong with the docker image itself, but a pretty long investigation showed that the problem was in the directory name. After I re-mounted the app to `/root/symfony` instead of just `/symfony` the error disappeared. Going further, I found the following pattern:

> When an app is in `/{vendor-name}`, composer's static autoloader stops working.

This is reproducible even without Docker:
```bash
# install symfony from http://symfony.com/download first
cd /tmp
symfony new seed

# our dir is /my-app (does not match any vendor name in composer.json)
sudo cp -R -p /tmp/seed /my-app
cd /my-app
composer install
# all OK

# our dir is /symfony (there exists a vendor called symfony in composer.json)
sudo cp -R -p /tmp/seed /symfony
cd /symfony
composer install
# Fatal error: Class 'Symfony\Component\ClassLoader\ClassCollectionLoader' not found
# in /symfony/vendor/sensio/distribution-bundle/Composer/ScriptHandler.php on line 282

# our dir is /doctrine (there exists a vendor called doctrine in composer.json)
sudo cp -R -p /tmp/seed /doctrine
cd /doctrine
composer install
# Fatal error: Class 'Doctrine\Common\Annotations\AnnotationRegistry' not found
# in /doctrine/app/autoload.php on line 11

# our dir is ~/symfony (same composer.json, but we are no longer in /)
cp -R -p /tmp/seed ~/symfony
cd ~/symfony
composer install
# all OK

sudo rm -rf /my-app
sudo rm -rf /symfony
sudo rm -rf /doctrine
sudo rm -rf ~/symfony
sudo rm -rf /tmp/seed
```

Something odd is happening with paths in `vendor/composer/autoload_static.php`. When we are in `/symfony`, there is this:

```php
public static $files = array (
    '92c8763cd6170fce6fcfe7e26b4e8c10' => __DIR__ . '/..' . __DIR__ . '/../..' . '/phpunit-bridge/bootstrap.php',
    '2c102faa651ef8ea5874edb585946bce' => __DIR__ . '/..' . '/swiftmailer/swiftmailer/lib/swift_required.php',
    '5255c38a0faeba867671b61dfda6d864' => __DIR__ . '/..' . '/paragonie/random_compat/lib/random.php',
    '023d27dca8066ef29e6739335ea73bad' => __DIR__ . '/..' . __DIR__ . '/../..' . '/polyfill-php70/bootstrap.php',
    'bd9634f2d41831496de0d3dfe4c94881' => __DIR__ . '/..' . __DIR__ . '/../..' . '/polyfill-php56/bootstrap.php',
    '0e6d7bf4a5811bfa5cf40c5ccd6fae6a' => __DIR__ . '/..' . __DIR__ . '/../..' . '/polyfill-mbstring/bootstrap.php',
    '6a47392539ca2329373e0d33e1dba053' => __DIR__ . '/..' . __DIR__ . '/../..' . '/polyfill-intl-icu/bootstrap.php',
    '32dcc8afd4335739640db7d200c1971d' => __DIR__ . '/..' . __DIR__ . '/../..' . '/polyfill-apcu/bootstrap.php',
);
```

Otherwise:
```php
public static $files = array (
    '92c8763cd6170fce6fcfe7e26b4e8c10' => __DIR__ . '/..' . '/symfony/phpunit-bridge/bootstrap.php',
    '2c102faa651ef8ea5874edb585946bce' => __DIR__ . '/..' . '/swiftmailer/swiftmailer/lib/swift_required.php',
    '5255c38a0faeba867671b61dfda6d864' => __DIR__ . '/..' . '/paragonie/random_compat/lib/random.php',
    '023d27dca8066ef29e6739335ea73bad' => __DIR__ . '/..' . '/symfony/polyfill-php70/bootstrap.php',
    'bd9634f2d41831496de0d3dfe4c94881' => __DIR__ . '/..' . '/symfony/polyfill-php56/bootstrap.php',
    '0e6d7bf4a5811bfa5cf40c5ccd6fae6a' => __DIR__ . '/..' . '/symfony/polyfill-mbstring/bootstrap.php',
    '6a47392539ca2329373e0d33e1dba053' => __DIR__ . '/..' . '/symfony/polyfill-intl-icu/bootstrap.php',
    '32dcc8afd4335739640db7d200c1971d' => __DIR__ . '/..' . '/symfony/polyfill-apcu/bootstrap.php',
);
```

When we are in `/doctrine`, there is:
```php
'Doctrine\\Instantiator\\' => 
array (
    0 => __DIR__ . '/..' . __DIR__ . '/../..' . '/instantiator/src/Doctrine/Instantiator',
),
'Doctrine\\Common\\Cache\\' => 
array (
    0 => __DIR__ . '/..' . __DIR__ . '/../..' . '/cache/lib/Doctrine/Common/Cache',
),
```
instead of:
```php
'Doctrine\\Instantiator\\' => 
array (
    0 => __DIR__ . '/..' . '/doctrine/instantiator/src/Doctrine/Instantiator',
),
'Doctrine\\Common\\Cache\\' => 
array (
    0 => __DIR__ . '/..' . '/doctrine/cache/lib/Doctrine/Common/Cache',
),
```

Avoiding static autoloader by manually modifying `vendor/composer/autoload_real.php` and then running `composer install --no-autoloader` returns the script to work. However, there is no way to avoid the issue without changing the directory of the project.

**Comment by alcohol** [comment]
That is a pretty funky edge-case..

Thanks for reporting it though. We will look into this.

**Comment by janpapenbrock** [comment]
Might be these two lines https://github.com/composer/composer/blob/a33b205561ddeb8c262a453dc64e158379c1ee5e/src/Composer/Autoload/AutoloadGenerator.php#L756-L757 

Probably the second is doing an unintended replacement in this case.

Would've loved to come up with a test for this, but that logic is somewhat buried and I couldn't get a grip on how the existing tests are structured.

**Comment by vaclavvanik** [comment]
@kachkaev Can you try `composer dump-autoload --optimize`?

**Comment by kachkaev** [comment]
@vaclavvanik I'm not on that project anymore, so won't be able to do this unfortunately. In any case, the result should be the same for everyone who repeats the steps in my first comment. The bug shows up in a brand new symfony project. Tested on Linux and OSX (php 5.6 I guess).

**Comment by kachkaev** [comment]
@vaclavvanik have you managed to reproduce the bug by a chance? My guess is that it should always appear after following the steps in my original commit, at least on *nix.

**Comment by everzet** [comment]
This bug is easily reproducible and seem to cause a lot of headache for people: https://github.com/Behat/Behat/issues/923

**Comment by alcohol** [comment]
It is not easily fixed unfortunately, but I am investigating further.

**Comment by alcohol** [comment]
@nicolas-grekas since you wrote this code; would you also care to have a peek? It is definitely related to the lines that @janpapenbrock linked to. I am not quite sure yet how to deal with the situation where a vendor directory name equals the name of the base app directory.

**Comment by alcohol** [comment]
In short, the issue is as follows:

Given that our project root is `/foo`
When we add package `foo/bar`

Things blow up because

``` php
$absoluteVendorPathCode = " '/foo/vendor/"
$absoluteAppBaseDirCode = " '/foo/"

$vendorPathCode = " __DIR__ . '/..' . '/"
$appBaseDirCode = " __DIR__ . '/../..' . '/"
```

So given an array of psr4 namespaces and directories:

``` php
"Foo\\Bar\\" => "/foo/vendor/foo/bar"
```

Those replace lines
``` php
$value = str_replace($absoluteVendorPathCode, $vendorPathCode, $value);
$value = str_replace($absoluteAppBaseDirCode, $appBaseDirCode, $value);
```

Give us
``` php
// 'Foo\\Bar\\' => 
// array (
//     0 => '/foo/vendor/foo/bar'
// )
//
// '/foo/vendor/foo/bar'
// 1st replace: ' __DIR__ . '/..' . '/foo/bar'
// 2nd replace: ' __DIR__ . '/..' . __DIR__ . '/../..' . '/bar'

"Foo\\Bar\\" => __DIR__ . '/..' . __DIR__ . '/../..' . '/bar'
```

**Comment by nicolas-grekas** [comment]
ACK, I'll look at this asap

**Comment by Seldaek** [comment]
Should be ok if we just replace it with a preg_replace anchored at the beginning of the string no?

**Comment by nicolas-grekas** [comment]
Can you please try #5633?

**Comment by Seldaek** [comment]
Fixed by #5633
